csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,319 @@
+ */
+// clang-format on
+
+#include <abstract_tensor.h>
+#include <abstract_tensor_schedule.h>
+#include <ir/internal_base_nodes.h>
+#include <iter_visitor.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+
+class AbstractTensorSchedule {
+ public:
+  static void apply(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph) {
+    AbstractTensorSchedule ats(abstract, concrete, graph);
+    ats.run();
+  }
+
+ private:
+  AbstractTensorSchedule(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph)
+      : abstract_(abstract), concrete_(concrete), graph_(graph) {}
+
+  void run() {
+    findNearestProducers();
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in concrete_ids_ and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    NVF_ERROR(
+        abs_id.is<ValGroupAndItsGraph>(),
+        "AbstractId must be a ValGroupAndItsGraph");
+    const ValGroupAndItsGraph& vgg = abs_id.as<ValGroupAndItsGraph>();
+    if (graph_ == nullptr) {
+      graph_ = vgg.graph;
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  //!
+  //! For example, suppose we have the following:
+  //!
+  //!   concrete_:
+  //!     root: iS0{i0} iS1{i1}
+  //!     loop: iS3{ceilDiv(i0 * i1, 16)} iS4{16}
+  //!     transforms:
+  //!       iS2 = merge(iS0, iS1)
+  //!       iS3, iS4 = split(iS2, 16)
+  //!   graph:
+  //!     ValGroup 0: 0 8
+  //!     ValGroup 1: 1 9
+  //!     ValGroup 2: 2 10
+  //!     ValGroup 3: 3
+  //!     ValGroup 4: 4
+  //!     ValGroup 5: 11
+  //!     ValGroup 6: 12
+  //!     ExprGroup 0: iS11{ceilDiv(i0 * i1, 32)}, iS12{32} = split(iS10, 32)
+  //!   abstract_:
+  //!     {{ValGroup 5, graph}, {ValGroup 6, graph}}
+  //!
+  //! Then findNearestProducers() work backward from the loop domain of
+  //! concrete_ (ValGroups 3 and 4) and will identify ValGroup 2 as a producer
+  //! to abstract_ and map it to the concrete domain iS2{i0}.
+  //!
+  //! Note that any dangling transforms will be discarded; in this example since
+  //! we work from iS2, we will discard the split that produced iS3 and iS4.
```

Comment:
Why `iS3` and `iS4` are considered dangling? Does this mean that if existing transforms of a TensorView could be altered?

Reply:
Yes, I use `setLoopDomain` instead of replaying transforms, and I use this function to find the starting point, which might not be IDs in the current loop domain. This means that there could be transforms in the current tensor that get erased.

Reply:
In the existing replay API, it isn't allowed to change existing transforms. Why do we need to change that here? Sounds a little dangerous to quietly discard transformations.

Reply:
You're right that this is different from the replay approach. My motivation is to get away from the statefulness of that approach. If we were to maintain transforms then that puts a burden on the schedulers to guarantee which transforms have already been applied to each tensor. Instead, this lets us do that with `AbstractTensor`s instead, which are easier to work with, can be copied, etc. However, I agree that silently discarding transforms is probably not ideal, unless we have a specific use case where it is helpful. So I think I will just assert that all the loop `IterDomain`s are found in the history of the `AbstractTensor`.

Reply:
Maybe if existing domains are set as loop or allocation domains, they won't be just discarded, right? 

In any case, I think this behavior would need to be more explicit.